nebula-logger/core/main/configuration/classes/LoggerCache.cls

```diff
@@ -0,0 +1,230 @@
+            }
+        }
+
+        /**
+         * @description Removes the specified `String` key from the organization cache
+         * @param  key  The `String` key to remove from the organization cache
+         */
+        public void remove(String key) {
+            getTransactionCache().remove(key);
+
+            if (LoggerParameter.USE_PLATFORM_CACHE == true) {
+                ORGANIZATION_CACHE_PARTITION.remove(key);
+            }
+        }
+    }
+
+    /**
+     * @description Manages any session-specific caching, using session platform cache & transaction cache
+     */
+    public class SessionCache implements Cacheable {
+        /**
+         * @description Indicates if the specified key has already been added to the cache
+         * @param  key The `String` key to check for within the session cache
+         * @return     The `Boolean` result that indicates if the specified key is contained in the cache
+         */
+        public Boolean contains(String key) {
+            if (LoggerParameter.USE_PLATFORM_CACHE == false || getTransactionCache().contains(key) || SESSION_CACHE_PARTITION.isAvailable() == false) {
+                return getTransactionCache().contains(key);
+            } else {
+                return SESSION_CACHE_PARTITION.contains(key);
+            }
+        }
+
+        /**
+         * @description Returns the cached value for the specified key, or `null` if
+         *              the specified key does not exist in the session cache
+         * @param  key  The `String` key to check for within the session cache
+         * @return      The cached value, or null if no cached value is found for the specified key
+         */
+        public Object get(String key) {
+            if (LoggerParameter.USE_PLATFORM_CACHE == false || getTransactionCache().contains(key) || SESSION_CACHE_PARTITION.isAvailable() == false) {
+                return getTransactionCache().get(key);
+            } else {
+                Object value = SESSION_CACHE_PARTITION.get(key);
+                getTransactionCache().put(key, value);
+                return value;
+            }
+        }
+
+        /**
+         * @description Adds the provided `Object` value to the current session's cache,
+         *              using the specified `String` key
+         * @param  key   The `String` key to add to the session cache
+         * @param  value The `Object` value to cache for the specified key
+         */
+        public void put(String key, Object value) {
+            getTransactionCache().put(key, value);
+
+            // Platform cache does not support storing null values
+            if (LoggerParameter.USE_PLATFORM_CACHE == true && value != null && SESSION_CACHE_PARTITION.isAvailable() == true) {
+                SESSION_CACHE_PARTITION.put(key, value, SESSION_CACHE_TTL_SECONDS, CACHE_VISIBILITY, CACHE_IS_IMMUTABLE);
```

Comment:
session cache supports nulls?

Reply:
makes sense - just wanted to make sure!

Reply:
I guess the question is - does it make sense for the transaction cache to be able to support nulls if the other two can't? Probably, unless you wanted to consolidate that null check. (Which maybe you can do anyway using the shared `Partition` variable?)

Reply:
Yeah, I think the transaction cache needs to support nulls, there are a few cached queries that return null in some scenarios (like when querying `AuthSession` during async jobs where there isn't an active `AuthSession`). I want to be able to cache the value even when null so that the queries don't constantly re-run - I guess there are other ways I could handle this, but having the transaction cache support null values seemed like the easiest & most consistent approach. 

And I think that duplicate null check is almost eliminated! I'm finishing up some of the consolidation to use the `Partition` class, it looks like that's going to work out really well.

Reply:
I think you wouldn't set it to null, you would remove the cache key, and then if you do a get, it should return null if the key doesn't exist I think.

Reply:
to @jongpie 's point, that would counteract the positive effect of transaction caching for some results. There's a big difference between `containsKey(key)` being false, and it being `true` with a null value; in the former, you don't know whether or not you've attempted to cache something, whereas with the latter you know you've cached something and the result was "empty"

Reply:
Yeah, I understand the difference but I wasn't sure if it mattered. If so, you could use a predefined value to represent a `null` state and just check for that when you retrieve the value. By the same token, the cache 'get' method probably shouldn't return null if the key is unset or the value is in the 'null' state. You may want to know that you tried to cache the value already and there is no sense attempting to generate it. This may require you to do something akin to this:
```
private static final String CACHE_IS_NULL = 'CACHE_IS_NULL';

...

Object oCache = LoggerCache.get(key);
if (oCache == null)
{
    // cache needs to be initialized
}
else if (LoggerCache.CASH_IS_NULL.equals(oCache)
{
    // cache was initialized, treat as null value though...
}
else
{
    // cast oCache and use as needed
}
```

Reply:
@rmccu thanks for the sample code! @jamessimone is correct that I don't want to remove the key when the value is `null`, but I do like the idea of a predefined value to represent `null` within platform cache - that seems like a clever way to handle it. I'll do a little prototyping with that approach to see how well it might work. I might still move forward with the current implementation for this PR as-is for now & implement the `CACHE_IS_NULL` later, but let me see what I can come up with in a quick prototype of this change.

Reply:
@rmccu this turned out to be a really nice & easy change, so I went ahead and committed the changes [here](https://github.com/jongpie/NebulaLogger/pull/394/commits/ae3db6fa7fa85bb213b8b988ff210e9578e86066). I was able to keep everything encapsulated within the `LoggerCache` class, so it handles the conversion to/from `null` to the substitute `String` value.

Classes that want to use caching can do something like this to know if a key has been cached/needs to be cached:

```java
public static SomeDataType getCachedValue() {
    String cacheKey = 'someKey';
    // Cache was initialized if contains(key) returns true - even if a null value was cached
    if (LoggerCache.getOrganizationCache().contains(cacheKey)) {
        // At this point, the code knows for sure that the specified key has been previously cached, 
        // so it just needs to cast & return the value. Internally, LoggerCache now handles dealing 
        // with nulls within Platform Cache - if a null value had been previously cached for
        // the specified key, then the Org Partition has the key stored with the substitute String value,
        // and LoggerCache.getOrganizationCache().get(cacheKey) returns null (not the substitute String value)

        // cast cached value and use as needed
        return (SomeDataType) LoggerCache.getOrganizationCache().get(cacheKey);
    } else {
        // cache needs to be initialized
        SomeDataType value; // TODO populate value
        LoggerCache.getOrganizationCache().put(cacheKey, value);
        return value;
    }
}
```

Reply:
Awesome! Glad it worked out smoothly. We did stuff like this at another company where we used the lower boundary values of an int (like negative 2 billion and change) to represent INT_NULL and INT_NOT_SET. It's not perfect, but it works.

I did have a side-question. Do you need to know the difference between null being initialized and not-initialized? For instance, if it is null because it is not initialized, that mean you need to go fetch the value to store and "initialize" the cache entry. Vs, when you know it is initialized and is null, you don't need to fetch.

I have had this problem with platform cache before. I think I punted and just took null to mean I need to initialize. This may incur additional cost if the value is null, but it was an edge case for me at the time.

Reply:
Doh! Just read the last part of 2nd message. We are good.

Reply:
Awesome, thanks @rmccu! I need to wrap up a little more testing & fix an unrelated failing test, but I'm hoping to finally get this PR merged & released today or tomorrow.